### PR TITLE
Fix test_thread_stats_usage fail on arm FVP

### LINKED ISSUE #46366: test_thread_stats_usage fail on arm64 fvp
```
INFO    - 1160/1160 fvp_base_revc_2xaemv8a    tests/kernel/usage/thread_runtime_stats/kernel.usage FAILED Failed (armfvp 3.568s)
...
START - test_thread_stats_usage
    Assertion failed at WEST_TOPDIR/zephyr-testing/tests/kernel/usage/thread_runtime_stats/src/test_thread_runtime_stats.c:517: test_thread_stats_usage: stats5.current_cycles < stats3.current_cycles is false
 FAIL - test_thread_stats_usage in 0.57 seconds
```

https://github.com/zephyrproject-rtos/zephyr-testing/runs/6745649272?check_suite_focus=true#step:11:1593

Note that this is a newly added test; see https://github.com/zephyrproject-rtos/zephyr/pull/45489

p.s. This also fails on `fvp_baser_aemv8r`; interestingly, it does not fail on `fvp_baser_aemv8r_aarch32`.

_Originally posted by @stephanosio in https://github.com/zephyrproject-rtos/zephyr/issues/45099#issuecomment-1146842381_

### PR DESCRIPTION
**Pull request by SgrrZhf**
This PR can fix https://github.com/zephyrproject-rtos/zephyr/issues/46366.

More detailed information can be seen in the code.

### CODE REVIEW

File: drivers/timer/arm_arch_timer.c [hunk 1]
```diff
@@ -98,6 +98,22 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
...
+	 * low-frequency clock.
+	 * For example, On FVP platform, When a timer compare interrupt occurs,
+	 * we got 530000 from cntvct_el0 in the arm_arch_timer_compare_isr
+	 * function and we may also get 530000 from cntvct_el0 here. This will
+	 * cause the req_cycle to round up to the wrong value and cause the
+	 * timer compare interrupt to occur earlier than expected.
+	 */
+	if (curr_cycle == last_cycle) {
+		curr_cycle += 1;
+	}
```
Comment (carlocaione) [inline drivers/timer/arm_arch_timer.c:115]: this feels a bit like cheating. Can we have this under an `ifdef` on the platform?

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: drivers/timer/arm_arch_timer.c (hunk 1 above)
Comment (povergoing) [reply drivers/timer/arm_arch_timer.c:115]: It's no problem to get this under an `ifdef`. For now, this is related to 4 FVP boards (or at least 2 SoCs, fvp aarch32 and fvp aarch64). 
But AFAIK, some hardware board also has this issue. Then the macro would be too much I think. And this logic won't affect the normal case and even could deal with the issue caused by a slower clock.
Well, it is ok for me to add the macro or not. Do you think it's better to not with the `ifdef`?

Comment (stephanosio) [inline drivers/timer/arm_arch_timer.c:115]: From https://github.com/zephyrproject-rtos/zephyr/issues/46366:

> p.s. This also fails on `fvp_baser_aemv8r`; interestingly, it does not fail on `fvp_baser_aemv8r_aarch32`.

So how can we explain this working properly on `fvp_baser_aemv8r_aarch32`?

Comment (povergoing) [reply drivers/timer/arm_arch_timer.c:115]: Or I may miss understanding your `ifdef`? Do you mean a feature macro, then the board with the issue select this?

Comment (carlocaione) [reply drivers/timer/arm_arch_timer.c:115]: > some hardware board also has this issue

Are there really boards out there as slow as FVP emulation?

> Do you mean a feature macro, then the board with the issue select this?

No, I meant an `ifdef` on the platform `#ifdef CONFIG_FVP_*` or equivalent.

Comment (SgrrZhf) [reply drivers/timer/arm_arch_timer.c:115]: Let me confirm it.

Comment (SgrrZhf) [reply drivers/timer/arm_arch_timer.c:115]: Got it.

**Comment by andyross** [comment]
I'm likewise suspicious about the "just add one to ensure monotonicity".  The way most timer drivers handle this kind of issue[1] is that they have a clamp somewhere of the minimum delta that can be set on the hardware to reliably produce an interrupt in the future.  Then they do the the (signed!) math and if the value is less they bump the interrupt to be scheduled forward by one tick. 

That scheme sounds like it might be a better fit here.  Or maybe the code is here and there's something else wrong?  I think this needs more thinking, basicallly.  This doesn't sound like a root cause fix to me.  "It should already work", basicallly.  Zephyr timer code should be robust vs. a slow timer polled twice in one cycle.

[1] Which is sort of a "race condition with real world time"

**Comment by SgrrZhf** [comment]
> I'm likewise suspicious about the "just add one to ensure monotonicity". The way most timer drivers handle this kind of issue[1] is that they have a clamp somewhere of the minimum delta that can be set on the hardware to reliably produce an interrupt in the future. Then they do the the (signed!) math and if the value is less they bump the interrupt to be scheduled forward by one tick.
> 
> That scheme sounds like it might be a better fit here. Or maybe the code is here and there's something else wrong? I think this needs more thinking, basicallly. This doesn't sound like a root cause fix to me. "It should already work", basicallly. Zephyr timer code should be robust vs. a slow timer polled twice in one cycle.
> 
> [1] Which is sort of a "race condition with real world time"

sorry, I may not fully understand your word. But I think I should give more information.

Let me explain something.

1. last_cycle is updated by arm_arch_timer_compare_isr.
2. Possibly, the CPU quickly runs to sys_clock_set_timeout and curr_cycle reads from the timer counter.
3. Assuming the timer counter is a bit of slow, so that right now we get the same value from the timer counter. That means curr_cycle is equal to last_cycle. Then when we calculate 'req_cycle', we will get the wrong value in line 120 and line 122(round up to next tick boundary). It's totally different when curr_cycle - last_cycle == 0 or curr_cycle - last_cycle >= 1. But I think it should has a same result with either curr_cycle - last_cycle == 0 or curr_cycle - last_cycle >= 1.
4. So in my initial fixes, if curr_cycle is equal to last_cycle, then add one to curr_cycle. And we will get calculate the right value for 'req_cycle'

**Comment by povergoing** [comment]
Based on the @SgrrZhf explanation, I think this is a kind of general issue. 
> That scheme sounds like it might be a better fit here. Or maybe the code is here and there's something else wrong? I think this needs more thinking, basicallly. This doesn't sound like a root cause fix to me. "It should already work", basicallly. **Zephyr timer code should be robust vs. a slow timer polled twice in one cycle**.

If we assume that the clock freq is low, then we probably get `curr_cycle - last_cycle` == 0, especially for the SMP (one core updated `last_cycle` in timer isr, and another core soon runs into `sys_clock_set_timeout`).
The issue is,
We have already misused the ticks by 1 in advance at the beginning of `sys_clock_set_timeout`:
https://github.com/zephyrproject-rtos/zephyr/blob/83e4f5dd5866f0f637d66e25a4d2ede2b835796c/drivers/timer/arm_arch_timer.c#L96-L97
but didn't add it back when `curr_cycle - last_cycle` == 0. We need `curr_cycle - last_cycle` at least to be `1` to get the right answer (round up to the next tick).
https://github.com/zephyrproject-rtos/zephyr/blob/83e4f5dd5866f0f637d66e25a4d2ede2b835796c/drivers/timer/arm_arch_timer.c#L104-L106
So the question is:
1) shall we consider the case with a lower clock?
2) if yes, do we have a better fix for this issue? Since the initial "add one" fixes sounds like cheating. I personally reckon this is not just for FVP.

May I ask for your suggestions? @andyross @carlocaione @stephanosio

**Comment by andyross** [comment]
I had a chance to take a look at this driver.  The root cause is that there's no clamping in the driver for a "minimum cycles" value.  Almost all hardware doesn't work if you try to set a comparator too soon in the future.  At the very least you want to avoid the aliasing case where you set the cycle that just expired (and thus have to wait for the counter to roll over).  Almost all Zephyr timer drivers have something like this.  Here's the one from the x64 HPET driver:

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/timer/hpet.c#L362

Note how there is a check vs. HPET_CMP_MIN_DELAY to avoid setting a delay too soon?  That's what you want to do here, because fixing the "1 cycle" case is (probably) not sufficient for your hardware, it's just the failure case you're seeing.

But also note that not all hardware needs this.  The apic_tsc driver (another x86 timer) doesn't have this, because the APIC is specified to be a proper comparator and not an equality test.  If you set a timeout in the "past" (in a 2's complement sense), it is guaranteed to latch the interrupt anyway and it will fire as soon as it is unmasked.  Maybe the intent was that the ARM timer was supposed to behave like this, but it has a glitch?

Anyway, it definitely seems like you stumbled on a condition like this, but instead of fixing it directly you're trying to freeload on the "round up to next tick" logic to do the increment for you.  That seems like a half solution (what if the tick rate is very high and one tick isn't enough?)

**Review by andyross — CHANGES_REQUESTED** [review]
(flip me over into the -1 camp, I'm 80% certain this is the wrong solution to a real problem)

**Comment by SgrrZhf** [comment]
I think you may misunderstand our issue. @andyross.

> The root cause is that there's no clamping in the driver for a "minimum cycles" value. Almost all hardware doesn't work if you try to set a comparator too soon in the future.

We do have a "minimum cycles" value limitation in the `arm_arch_timer` driver.
https://github.com/zephyrproject-rtos/zephyr/blob/662cea46388b714d1546affea53d56c3d6562b38/drivers/timer/arm_arch_timer.c#L108

> Note how there is a check vs. HPET_CMP_MIN_DELAY to avoid setting a delay too soon? That's what you want to do here, because fixing the "1 cycle" case is (probably) not sufficient for your hardware, it's just the failure case you're seeing.

No, this issue isn't about setting a delay too soon, this is about too soon to set a delay. Below is a timeline, each integer in the timeline represents the value of the timer cycle.
`0--------A(530000)-------B(530001)--------------------------C(620000) -----------D(630000)--------->  cycle`

Let's consider a situation. _CYC_PER_TICK = 100000_
1. At point A, we enter the timer ISR and the variable `last_cycle` will update to 530000.
2. Then we start to set the next timeout in ISR, assuming we want to set the next timeout after 10 ticks. It means the parameter of the function `sys_clock_set_timeout` is `(10, false)`.
3. What if procedure 2 occurs after point A and before point B. In this situation, _last_cycle=530000_, _curr_cycle=530000_
    - We expect the variable `req_cycle` to equal to '100000' and the next timeout to occur at point D. 
    - But the `req_cycle` will equal to '90000' and the next timeout will occur at point C.
https://github.com/zephyrproject-rtos/zephyr/blob/83e4f5dd5866f0f637d66e25a4d2ede2b835796c/drivers/timer/arm_arch_timer.c#L104-L106
It's similar to https://github.com/zephyrproject-rtos/zephyr/blob/4946a15f157917bdaffdd1fddc4e875d8cc75e7f/drivers/timer/hpet.c#L352-L359
    - If we add 1 to the variable `curr_cycle` when procedure 2 occurs after point A and before point B, we will get a good result at last and the next timeout will occur at point D.
4. Whatever point A equals, The issue will always occur when procedure 2 occurs after point A and before point B if we don't change anything.

**Comment by andyross** [comment]
OK... I think I'm getting it.  Let me rephrase and see if this matches your understanding and/or explains things better to the other folks here who were confused:

1. The kernel is calling sys_clock_set_timeout() with a ticks argument of 0 or 1 to indicate "set a timeout at the next tick"
2. The hardware, because of either a slow clock or very fast handling, has not advanced and so the current cycle counter is still equal to the tick-aligned `last_cycle` value for which it was set.
3. The math assumes the scheduled cycle expiration is in the future and so "round up" results in a noop.

So you need to make sure that "round up" actually **advanced forward** to the next tick.  We don't want "rounding" at all.

Basically I've flipped and think you're fix is broadly correct and in the right spot, but:

1.  The "add one always" algorithm isn't really correct and will give you a symmetric 1-cycle aliasing problem at the **end** of a tick (where the resulting interrupt will be one tick late).   I wouldn't be surprised if there was an obscure test failure lurking here either.  You probably want to implement a clamping operation instead such that the "advance forward to" semantics are honored.

2. This is a pretty subtle edge case and needs a good comment to explain what's going on

3. This bug probably exists (even if it's not detectable due to cycle rates and interrupt latency) on every other driver in the tree.  We probably want to file a tracking bug to see that they get audited and maybe come up with a framework for them all to use to unify all the extremely similar cycle logic.

**Comment by andyross** [comment]
(And thanks for your patience as I puzzle this out.)

**Comment by SgrrZhf** [comment]
No problem, I'm happy to contribute to the Zephyr project. @andyross 
It seems that there is a lot of work to fix this bug.
May I ask, what could I do to this PR or is there anything that I can do to help you resolve this bug?

**Comment by povergoing** [comment]
What about
`req_cycle += max(curr_cycle - last_cycle, 1) + (CYC_PER_TICK - 1); `
instead of adding `1` to `curr_cycle`

**Comment by andyross** [comment]
Tell you what, replace the comment with one like this (feel free to rephrase) that explains the circumstance and impact and I'll +1.  I'll file the tracking bug.

```c
/* Fast hardware with slow timer hardware can trigger and enter an
 * interrupt and reach this spot before the counter has advanced.
 * That defeats the "round up" logic such that we end up scheduling
 * timeouts a tick too soon (e.g. if the kernel requests an interrupt
 * at the "next" tick, we would end up computing a comparator value
 * representing the current time!).  Offset to correct.
 */
```

I think I'd still prefer to see the code expressed in the sense of "compute the expiration time from first principles via 'advance' instead of 'round up'" (i.e. with a MIN() macro to set the floor instead of hacking in an offset) but realistically that's just going to generate functionally identical code and isn't worth arguing about.

**Review by andyross — DISMISSED** [review]
Thanks for your patience

**Review by stephanosio — DISMISSED** [review]

**Review by povergoing — DISMISSED** [review]

**Review by povergoing — APPROVED** [review]
refresh +1

File: drivers/timer/arm_arch_timer.c (hunk 1 above)
Comment (SgrrZhf) [reply drivers/timer/arm_arch_timer.c]: This conversation may out of date. Could I just click `resolve conversation` directly? @carlocaione @povergoing

**Comment by SgrrZhf** [comment]
My branch's commit history isn't clean cause I clicked the `Update branch` button by mistake. So I rebased and force-pushed again.😅

**Review by carlocaione — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]
